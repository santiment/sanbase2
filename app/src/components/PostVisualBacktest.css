.post-visual-backtest {
  line-height: 16px;
}

.post-visual-backtest__info {
  margin-right: 4px;
  margin-bottom: 1em;
  display: flex;
}

@media (min-width: 768px) {
  .post-visual-backtest {
    display: flex;
  }

  .post-visual-backtest__info {
    flex-direction: column;
  }
}
