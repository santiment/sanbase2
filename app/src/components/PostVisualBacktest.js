import React from 'react'
import PropTypes from 'prop-types'
import { Message } from 'semantic-ui-react'
import { graphql } from 'react-apollo'
import { compose, withProps } from 'recompose'
import { HistoryPriceGQL } from './../pages/Detailed/DetailedGQL'
import PercentChanges from './PercentChanges'

const getChanges = (start, last, prop = 'priceUsd') =>
  (last[`${prop}`] - start[`${prop}`]) / start[`${prop}`] * 100

const isTotalMarket = ticker => ticker === 'Crypto Market'

const propTypes = {
  ticker: PropTypes.string.isRequired,
  history: PropTypes.object
}

const PostVisualBacktest = ({ ticker, change, changeProp }) => {
  if (!change) return ''
  return (
    <Message>
      {changeProp} changes after publication
      {change && <PercentChanges changes={change} />}
    </Message>
  )
}

const enhance = compose(
  graphql(HistoryPriceGQL, {
    name: 'history',
    options: ({ ticker, from }) => {
      return {
        skip: !ticker || !from,
        errorPolicy: 'all',
        variables: {
          from,
          ticker: isTotalMarket(ticker) ? 'TOTAL_MARKET' : ticker
        }
      }
    }
  }),
  withProps(({ ticker, history: { historyPrice = [] } }) => {
    if (!historyPrice) return
    const start = historyPrice[0]
    const last = historyPrice[historyPrice.length - 1]
    if (!start || !last) return
    const changeProp = isTotalMarket(ticker) ? 'Total marketcap' : 'Prices'
    return {
      change: getChanges(
        start,
        last,
        isTotalMarket(ticker) ? 'marketcap' : 'priceUsd'
      ),
      changeProp
    }
  })
)

PostVisualBacktest.propTypes = propTypes

PostVisualBacktest.defaultProps = {
  history: {
    historyPrice: []
  }
}

export default enhance(PostVisualBacktest)
