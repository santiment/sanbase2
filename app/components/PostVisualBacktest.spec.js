/* eslint-env jest */
import React from 'react'
import { shallow, mount, render } from 'enzyme'
import toJson from 'enzyme-to-json'
import { PostVisualBacktest } from './PostVisualBacktest'

describe('PostVisualBacktest component', () => {
  it('(smoke) it should render correctly', () => {
    const component = render(<PostVisualBacktest ticker='SAN' from='aksdfjh' />)
    expect(toJson(component)).toMatchSnapshot()
    expect(component.html()).toEqual(null)
  })

  it('SAN icon should render correctly', () => {
    const component = render(
      <PostVisualBacktest
        ticker='SAN'
        change={25}
        changeProp='priceUsd'
        from='aksdfjh'
      />
    )
    expect(toJson(component)).toMatchSnapshot()
  })
})
