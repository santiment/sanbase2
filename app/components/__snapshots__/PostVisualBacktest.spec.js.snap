// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostVisualBacktest component (smoke) it should render correctly 1`] = `null`;

exports[`PostVisualBacktest component SAN icon should render correctly 1`] = `
<div
  class="post-visual-backtest"
>
  <div
    class="post-visual-backtest__info"
  >
    <div
      class="post-visual-backtest__changes"
    >
      SAN priceUsd since publication
    </div>
    <div
      class="percent-changes percent-changes--positive"
    >
      <i
        class="fa fa-caret-up"
      />
      Â 25.00%
    </div>
  </div>
  <div
    class="PostVisualBacktestChart"
  >
    <div
      class="recharts-responsive-container"
      style="width:100%;height:100%"
    >
      <div
        style="position:absolute;width:0;height:0;visibility:hidden;display:none"
      />
    </div>
  </div>
</div>
`;
